k2/csrc/cuda/debug.cuh

```diff
@@ -0,0 +1,169 @@
+// k2/csrc/cuda/errors.h
+
+// Copyright (c) 2020, Xiaomi Corporation ( authors: Meixu Song )
+
+// See LICENSE for clarification regarding multiple authors
+
+#ifndef K2_CSRC_CUDA_ERRORS_H_
+#define K2_CSRC_CUDA_ERRORS_H_
+
+#include <cstdio>
+
+#include "k2/csrc/cuda/arch.h"
+
+namespace k2 {
+
+// In debug mode and include this header,
+// define K2_MAKE_ERROR_CHECK to turn on error checking
+#if (!defined(NDEBUG) && !defined(K2_MAKE_ERROR_CHECK))
+  #define K2_MAKE_ERROR_CHECK
+#endif
+
+/**
+ * \brief A static assertion
+ * \param exp the compile-time boolean expression that must be true
+ * \param msg an error message if exp is false
+ *
+ * `static_assert` is supported by both of host and device.
+ */
+#ifndef K2_STATIC_ASSERT
+  #define K2_STATIC_ASSERT(exp, msg) static_assert(exp, msg)
+#endif
+
+/**
+ * \brief This is a error checking function, with context information.
+ * If K2_MAKE_ERROR_CHECK is defined and error is not cudaSuccess,
+ * the corresponding error message is printed to:
+ *   - host: stderr
+ *   - device: stdout in device
+ * along with the supplied source context.
+ * It's used to make other macros convenient.
+ *
+ * \return The CUDA error.
+ */
+__host__ __device__ __forceinline__ cudaError_t K2_CUDA_DEBUG(
+    cudaError_t error,
+    const char *filename,
+    int line,
+    bool abort = true) {
+#ifdef K2_MAKE_ERROR_CHECK
+  (void)filename;
+  (void)line;
+  if (cudaSuccess != error) {
+  #if (K2_PTX_ARCH == 0)
+    fprintf(stderr, "CUDA error ID=%d, NAME=%s, [%s, %d]: %s\n",
+            error, cudaGetErrorName(error),
+            filename, line,
+            cudaGetErrorString(error));
+    fflush(stderr);
+  #elif (K2_PTX_ARCH >= 200)
+    printf("CUDA error ID=%d, NAME=%s, [block (%d,%d,%d) thread (%d,%d,%d), %s, %d]: %s\n",
+           error, cudaGetErrorName(error),
+           blockIdx.z, blockIdx.y, blockIdx.x, threadIdx.z, threadIdx.y, threadIdx.x,
```

Comment:
it's strange that the order is `z, y, x` instead of `x, y, z`. and can we add kernel name here?

Reply:
BTW, where this function is supposed to be called? It seems it will be called with return value (`cudaError_t error`) from some cuda api, if it's not a kernel, how can we get `threadIdx` and `blockIdx` here?

Reply:
Can you please give an example of how this would be called? 
I want to see what kind of code it's intended for.
I see that there are examples in CUDA API's of device functions that return cudaError_t.  But they're probably not common... if this is not intended to be used frequently, please clarify.

Reply:
> it's strange that the order is `z, y, x` instead of `x, y, z`. and can we add kernel name here?

As the CUDA thread/block indexes are column-major distributed. I.e. they are virtual concepts for developers. For the indexing of them, if it's print like `(A, B, C)`, then it implies `[a][b][c]` as a way to index them. Thus, according to the column-major, `a=z, b=y, c=x`. However, I can't found a convention to log them. So, anyway is ok to me.

Reply:
> Can you please give an example of how this would be called?
> I want to see what kind of code it's intended for.
> I see that there are examples in CUDA API's of device functions that return cudaError_t. But they're probably not common... if this is not intended to be used frequently, please clarify.

OK. I am writing gtests/examples.

Reply:
> As the CUDA thread/block indexes are column-major distributed

IIRC，the index is row-major...anyway, I think (x,y,z) is more straightforward to most people, this is also used by most of cuda examples

Reply:
> The K2_PTX_ARCH is used to solve this. During compiling, as to the host part, it's 0; as to the kernel part, it's not.

Sorry, I don't understand it clearly, Can you explain it with some code? I mean, How can you know it's host code or device code and annotate the code with macro `K2_PTX_ARCH` ? I did't find the related code in the PR.

Reply:
> > The K2_PTX_ARCH is used to solve this. During compiling, as to the host part, it's 0; as to the kernel part, it's not.
> 
> Sorry, I don't understand it clearly, Can you explain it with some code? I mean, How can you know it's host code or device code and annotate the code with macro `K2_PTX_ARCH` ? I did't find the related code in the PR.

It's defined in `cuda/arch.h` with comments.